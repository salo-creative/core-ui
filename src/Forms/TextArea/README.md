# TextArea

This component is designed to be used as a textarea element. Pass a value to it to fill it in.

It can take a label, helper text, an error warning, etc. It can be set to a disabled state as well.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import TextArea from '@salo/core-ui/Forms/TextArea';
```

Implement as follows

```javascript
<TextArea 
  value="This text would show up inside the TextArea field."
  onChange={ ({ e, value }) => console.log({ e, value }) }
  onBlur={ ({ e, value }) => console.log({ e, value }) }
/>
```

For full prop types and usage see storybook info/knobs