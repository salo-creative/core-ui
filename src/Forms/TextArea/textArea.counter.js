import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

const Counter = ({ countTo, value }) => {
  const count = countTo - get(value, 'length') || 0;
  
  return (
    <p className='salo-textarea-counter'>
      { count } <span className='salo-textarea-counter-text'>characters</span>
    </p>
  );
};

Counter.propTypes = {
  countTo: PropTypes.number.isRequired,
  value: PropTypes.string.isRequired
};

export default Counter;