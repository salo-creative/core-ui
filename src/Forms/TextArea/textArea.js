import React from 'react';
import PropTypes from 'prop-types';

import { TextAreaWrapper, Field } from './textArea.styles';
import Label from '../components/Label';
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';
import Counter from './textArea.counter';

// HELPERS & CONSTANTS
import { colours } from '../../helpers/colours';

const TextArea = props => {
  const {
    background,
    className,
    countTo,
    disabled,
    error,
    errorMessage,
    helperText,
    label,
    margin,
    max,
    min,
    name,
    onBlur,
    onChange,
    onFocus,
    placeholder,
    required,
    rows,
    size,
    value
  } = props;

  return (
    <TextAreaWrapper margin={ margin } className={ `salo-textarea ${ className }` }>
      <Label
        className='salo-textarea__label'
        error={ error }
        label={ label }
        name={ name }
        required={ required }
        size={ size }
      />
      <Field
        aria-required={ required.toString() }
        aria-invalid={ error.toString() }
        background={ background }
        className='salo-textarea__field'
        disabled={ disabled }
        id={ name }
        maxLength={ max }
        minLength={ min }
        onBlur={ e => onBlur({ e, value: e.target.value }) }
        onChange={ e => onChange({ e, value: e.target.value }) }
        onFocus={ e => onFocus({ e, value: e.target.value }) }
        placeholder={ placeholder }
        rows={ rows }
        value={ value }
      />
      { countTo && (
        <Counter
          countTo={ countTo }
          value={ value }
        />
      ) }
      <ErrorText
        className='salo-textarea__error'
        disabled={ disabled }
        error={ error }
        errorMessage={ errorMessage }
        size={ size }
      />
      <HelperText
        className='salo-textarea__helper'
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        size={ size }
      />
    </TextAreaWrapper>
  );
};

TextArea.defaultProps = {
  background: colours.grey,
  className: '',
  countTo: null,
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  helperText: '',
  label: '',
  margin: '0 0 2rem',
  max: null,
  min: null,
  onBlur: () => null,
  onChange: () => null,
  onFocus: () => null,
  placeholder: '',
  required: false,
  rows: 3,
  size: 'M',
  value: ''
};

TextArea.propTypes = {
  background: PropTypes.string,
  className: PropTypes.string,
  countTo: PropTypes.number,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  helperText: PropTypes.string,
  label: PropTypes.string,
  margin: PropTypes.string,
  max: PropTypes.number,
  min: PropTypes.number,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  rows: PropTypes.number,
  size: PropTypes.oneOf(['L', 'M']),
  value: PropTypes.string
};

export default TextArea;