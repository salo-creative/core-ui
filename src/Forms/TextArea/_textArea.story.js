import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, number, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// Tests.
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// Featured Component.
import { TextArea } from '../../index';

// README
import README from './README.md';

// Story logic.
const stories = storiesOf('Forms | TextArea', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['textarea'] });

stories.add(
  'Basic',
  (() => {
    // Initialize store.
    const store = new Store({ value: 'Some text would sure look nice here.' });

    // Knobs.
    const label = text('Label', 'TextArea Label');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const rows = number('Rows', 3);
    const helperText = text('Helper text', 'Helper text');
    const size = select('size', ['L', 'M'], 'M');
    return (
      <RenderWithProps store={ store }>
        <TextArea
          name='story'
          disabled={ disabled }
          error={ error }
          label={ label }
          required={ required }
          helperText={ helperText }
          value={ store.value }
          onChange={ ({ value }) => store.set({ value }) }
          rows={ rows }
          size={ size }
        />
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);

stories.add(
  'Counter',
  (() => {
    // Initialize store.
    const store = new Store({ value: 'Some text would sure look nice here.' });

    // Knobs.
    const label = text('Label', 'TextArea Label');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const rows = number('Rows', 3);
    const maxSize = number('Max characters', 200);
    const helperText = text('Helper text', 'Helper text');
    const size = select('size', ['L', 'M'], 'M');
    return (
      <RenderWithProps store={ store }>
        <TextArea
          countTo={ maxSize }
          disabled={ disabled }
          error={ error }
          helperText={ helperText }
          label={ label }
          max={ maxSize }
          name='story'
          onChange={ ({ value }) => store.set({ value }) }
          required={ required }
          rows={ rows }
          size={ size }
          value={ store.value }
        />
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);