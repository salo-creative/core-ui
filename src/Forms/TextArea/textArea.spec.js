import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import TextArea from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <TextArea />
  );
  expect(container.firstChild).toMatchSnapshot();
});