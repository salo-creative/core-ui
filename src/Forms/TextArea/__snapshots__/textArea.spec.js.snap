// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 0 2rem;
}

.c1 {
  margin: 0;
  padding: 1rem;
  background: #fafcff;
  border: 1px solid;
  border-color: rgba(0,172,237,0.5);
  -webkit-transition: border 0.2s linear;
  transition: border 0.2s linear;
  width: 100%;
  border-radius: 0.4rem;
  font-size: 1.4rem;
}

.c1:focus,
.c1:active {
  border-color: #00ACED;
}

.c1[disabled] {
  cursor: not-allowed;
  opacity: 0.75;
}

<div
  class="salo-textarea  c0"
>
  <textarea
    aria-invalid="false"
    aria-required="false"
    class="salo-textarea__field c1"
    placeholder=""
    rows="3"
  />
</div>
`;
